File: app/src/main/java/com/faendir/lightning_launcher/multitool/settings/HasPlayerEntries.java
package com.faendir.lightning_launcher.multitool.settings;

import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import androidx.annotation.NonNull;
import java9.util.stream.StreamSupport;

import java.util.List;

/**
 * @author lukas
 * @since 21.06.18
 */
public interface HasPlayerEntries {
    void setEntries(CharSequence[] entries);

    void setEntryValues(CharSequence[] entryValues);

    default void discoverPlayers(@NonNull PackageManager pm) {
        List<ResolveInfo> infos = pm.queryBroadcastReceivers(new Intent(Intent.ACTION_MEDIA_BUTTON), 0);
        setEntries(StreamSupport.stream(infos).map(info -> info.activityInfo.loadLabel(pm).toString()).toArray(String[]::new));
        setEntryValues(StreamSupport.stream(infos).map(info -> info.activityInfo.packageName).toArray(String[]::new));
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/settings/HasPlayerEntries.kt
package com.faendir.lightning_launcher.multitool.settings

import android.content.Intent
import android.content.pm.PackageManager

/**
 * @author lukas
 * @since 21.06.18
 */
interface HasPlayerEntries {
    fun setEntries(entries: Array<CharSequence>)

    fun setEntryValues(entryValues: Array<CharSequence>)

    fun discoverPlayers(pm: PackageManager) {
        val infos = pm.queryBroadcastReceivers(Intent(Intent.ACTION_MEDIA_BUTTON), 0)
        setEntries(infos.map { info -> info.activityInfo.loadLabel(pm).toString() }.toTypedArray())
        setEntryValues(infos.map { info -> info.activityInfo.packageName }.toTypedArray())
    }
}
